File: ios/Sources/NoiseProtocol/State/SymmetricState.swift
import Foundation

/// Wraps CipherState with chaining key and handshake hash.
public final class SymmetricState: @unchecked Sendable {
    internal let suite: CipherSuite
    private var ck: Data
    private var h: Data
    private let cipherState: CipherState
    /// True if the handshake uses PSK mode (modifies "e" token processing).
    internal var hasPSK: Bool = false

    public init(suite: CipherSuite = .noise_25519_ChaChaPoly_SHA256) {
        self.suite = suite
        ck = Data()
        h = Data()
        cipherState = CipherState(suite: suite)
    }

    public func initializeSymmetric(protocolName: String) {
        let nameData = Data(protocolName.utf8)
        if nameData.count <= suite.hashlen {
            h = nameData + Data(repeating: 0, count: suite.hashlen - nameData.count)
        } else {
            h = suite.hash(nameData)
        }
        ck = h
        cipherState.initializeKey(nil)
    }

    public func mixKey(_ inputKeyMaterial: Data) {
        let outputs = suite.hkdf(chainingKey: ck, inputKeyMaterial: inputKeyMaterial, numOutputs: 2)
        ck = outputs[0]
        var tempK = outputs[1]
        if suite.hashlen > 32 {
            tempK = Data(tempK.prefix(32))
        }
        cipherState.initializeKey(tempK)
    }

    public func mixHash(_ data: Data) {
        h = suite.hash(h + data)
    }

    /// Used for PSK mode. Mixes key material into ck, h, and cipher key.
    public func mixKeyAndHash(_ inputKeyMaterial: Data) {
        let outputs = suite.hkdf(chainingKey: ck, inputKeyMaterial: inputKeyMaterial, numOutputs: 3)
        ck = outputs[0]
        mixHash(outputs[1])
        var tempK = outputs[2]
        if suite.hashlen > 32 {
            tempK = Data(tempK.prefix(32))
        }
        cipherState.initializeKey(tempK)
    }

    public func getHandshakeHash() -> Data { h }

    /// Encrypts plaintext using h as AD, then mixes ciphertext into h.
    public func encryptAndHash(_ plaintext: Data) throws -> Data {
        let ciphertext = try cipherState.encryptWithAd(h, plaintext: plaintext)
        mixHash(ciphertext)
        return ciphertext
    }

    /// Decrypts ciphertext using h as AD, then mixes ciphertext into h.
    public func decryptAndHash(_ ciphertext: Data) throws -> Data {
        let plaintext = try cipherState.decryptWithAd(h, ciphertext: ciphertext)
        mixHash(ciphertext)
        return plaintext
    }

    /// Splits the symmetric state into two CipherStates for transport.
    /// c1 encrypts initiator→responder, c2 encrypts responder→initiator.
    public func split() -> (CipherState, CipherState) {
        let outputs = suite.hkdf(chainingKey: ck, inputKeyMaterial: Data(), numOutputs: 2)
        var tempK1 = outputs[0]
        var tempK2 = outputs[1]
        if suite.hashlen > 32 {
            tempK1 = Data(tempK1.prefix(32))
            tempK2 = Data(tempK2.prefix(32))
        }
        let c1 = CipherState(suite: suite)
        let c2 = CipherState(suite: suite)
        c1.initializeKey(tempK1)
        c2.initializeKey(tempK2)
        return (c1, c2)
    }

    internal var hasKey: Bool { cipherState.hasKey }
}

File: android/lib/src/main/kotlin/com/noise/protocol/state/SymmetricState.kt
package com.noise.protocol.state

import com.noise.protocol.crypto.CipherSuite

/**
 * Wraps CipherState with chaining key and handshake hash.
 */
class SymmetricState(private val suite: CipherSuite = CipherSuite.NOISE_25519_CHACHAPOLY_SHA256) {
    private var ck: ByteArray = ByteArray(0)
    private var h: ByteArray = ByteArray(0)
    private val cipherState = CipherState(suite)

    /** True if the handshake uses PSK mode (modifies "e" token processing). */
    var hasPSK: Boolean = false

    fun initializeSymmetric(protocolName: String) {
        val nameBytes = protocolName.toByteArray(Charsets.UTF_8)
        h = if (nameBytes.size <= suite.hashlen) {
            nameBytes + ByteArray(suite.hashlen - nameBytes.size)
        } else {
            suite.hash(nameBytes)
        }
        ck = h.copyOf()
        cipherState.initializeKey(null)
    }

    fun mixKey(inputKeyMaterial: ByteArray) {
        val outputs = suite.hkdf(ck, inputKeyMaterial, 2)
        ck = outputs[0]
        var tempK = outputs[1]
        if (suite.hashlen > 32) {
            tempK = tempK.copyOfRange(0, 32)
        }
        cipherState.initializeKey(tempK)
    }

    fun mixHash(data: ByteArray) {
        h = suite.hash(h + data)
    }

    /** Used for PSK mode. Mixes key material into ck, h, and cipher key. */
    fun mixKeyAndHash(inputKeyMaterial: ByteArray) {
        val outputs = suite.hkdf(ck, inputKeyMaterial, 3)
        ck = outputs[0]
        mixHash(outputs[1])
        var tempK = outputs[2]
        if (suite.hashlen > 32) {
            tempK = tempK.copyOfRange(0, 32)
        }
        cipherState.initializeKey(tempK)
    }

    fun getHandshakeHash(): ByteArray = h.copyOf()

    /** Encrypts plaintext using h as AD, then mixes ciphertext into h. */
    fun encryptAndHash(plaintext: ByteArray): ByteArray {
        val ciphertext = cipherState.encryptWithAd(h, plaintext)
        mixHash(ciphertext)
        return ciphertext
    }

    /** Decrypts ciphertext using h as AD, then mixes ciphertext into h. */
    fun decryptAndHash(ciphertext: ByteArray): ByteArray {
        val plaintext = cipherState.decryptWithAd(h, ciphertext)
        mixHash(ciphertext)
        return plaintext
    }

    /**
     * Splits into two CipherStates for transport.
     * c1 encrypts initiator→responder, c2 encrypts responder→initiator.
     */
    fun split(): Pair<CipherState, CipherState> {
        val outputs = suite.hkdf(ck, ByteArray(0), 2)
        var tempK1 = outputs[0]
        var tempK2 = outputs[1]
        if (suite.hashlen > 32) {
            tempK1 = tempK1.copyOfRange(0, 32)
            tempK2 = tempK2.copyOfRange(0, 32)
        }
        val c1 = CipherState(suite)
        val c2 = CipherState(suite)
        c1.initializeKey(tempK1)
        c2.initializeKey(tempK2)
        return c1 to c2
    }

    val hasKey: Boolean get() = cipherState.hasKey
}
